### PR TITLE
Typescript migration

### PR DESCRIPTION
**Pull request by wtrocki**
Creating branch for collaboration and to not block the work. 
Done early migration in order to be able to compile the changes. 
There is much more work needed. The way we can collaborate on this is to take one file at the time from now since most of the files might be working fine.

Migration was smooth. 
I have spent most of the time rebasing/resolving conflicts with the changes on master.
Migration exposed couple problems. I have marked them as TODO in the code.

## Progress

Mark file that was processed. After all, build should be passing

- [x] Source code moved to typescript

Processed so far (100%)

├── KStorage.ts 
├── KafkaFactory.ts
├── KafkaStreams.ts
├── StorageMerger.ts
├── actions
│   ├── KeyCount.ts
│   ├── LastState.ts
│   ├── Max.ts
│   ├── Min.ts
│   ├── Sum.ts
│   ├── Window.ts
│   └── index.ts
├── client
│   ├── JSKafkaClient.ts
│   ├── KafkaClient.ts
│   └── NativeKafkaClient.ts
├── dsl
│   ├── KStream.ts
│   ├── KTable.ts
│   └── StreamDSL.ts
├── messageProduceHandle.ts
└── produceTypes.ts

- [x] Update unit tests and move them to typescript
@rob3000 WDYT about doing this?
- [ ] Update documentation

### CODE REVIEW

File: src/lib/KafkaStreams.ts [hunk 1]
```diff
@@ -1,17 +1,22 @@
...
+import KStream from './dsl/KStream.js';
+import KTable from './dsl/KTable.js';
+import KStorage from './KStorage.js';
 
 /**
  * Stream object factory
  * inhabits EventEmitter(events)
  */
 class KafkaStreams extends EventEmitter {
+	public config: any;
```
Comment (wtrocki) [inline src/lib/KafkaStreams.ts:14]: Kept them public for the moment.

**Comment by wtrocki** [comment]
Looks like things failing on CI but passing locally. Need to look better. 

@rob3000 any comments?

File: src/lib/client/JSKafkaClient.ts [hunk 2]
```diff
@@ -1,14 +1,30 @@
 "use strict";
 
-const uuid = require("uuid");
-const { Kafka, PartitionDrainer, Drainer, Publisher } = require("sinek");
-const debug = require("debug")("kafka-streams:jsclient");
-
-const KafkaClient = require("./KafkaClient.js");
+import uuid from 'uuid';
```
Comment (rob3000) [inline src/lib/client/JSKafkaClient.ts:3]: Actions failing due to: Drainer expecting 5 options, even though some of them should be optional: https://github.com/nodefluent/node-sinek/blob/16395fefbe37cca09d751dafd421040bd4834288/index.d.ts#L356

https://github.com/nodefluent/node-sinek/blob/16395fefbe37cca09d751dafd421040bd4834288/lib/kafka/Drainer.js#L12

Comment (rob3000) [reply src/lib/client/JSKafkaClient.ts:3]: Have fixed this as part of: https://github.com/nodefluent/kafka-streams/pull/166

**Comment by rob3000** [comment]
Created a new branch to merge into this PR: https://github.com/nodefluent/kafka-streams/pull/166

This PR looks good to me and will approve. I have fixed some indentation and a few other items which i have added to the above PR. This one can be merged :)

File: src/index.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
+"use strict";
```
Comment (rob3000) [inline src/index.ts:1]: Have removed this and added as part of the ts build.

File: src/index.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+import  KStream  from "./lib/dsl/KStream"
+import  KTable  from "./lib/dsl/KTable"
+import  KafkaFactory  from "./lib/KafkaFactory"
+import  KafkaStreams  from "./lib/KafkaStreams"
+import  KStorage  from "./lib/KStorage.js"
+import  KafkaClient  from "./lib/client/KafkaClient"
+
+module.exports = {
```
Comment (rob3000) [inline src/index.ts:10]: I have fixed this up to use named exports as normal exports in https://github.com/nodefluent/kafka-streams/pull/166

File: src/lib/KafkaFactory.ts [hunk 5]
```diff
@@ -39,4 +41,4 @@ class KafkaFactory {
   }
 }
 
-module.exports = KafkaFactory;
+export default KafkaFactory;
```
Comment (rob3000) [inline src/lib/KafkaFactory.ts:44]: I've also moved the export notation to where the class is defined in: https://github.com/nodefluent/kafka-streams/pull/166

**Review by rob3000 — APPROVED** [review]

File: src/index.ts (hunk 3 above)
Comment (wtrocki) [reply src/index.ts:1]: yep.. I forgot to remove that from every file.

File: src/index.ts (hunk 4 above)
Comment (wtrocki) [reply src/index.ts:10]: Oh.. I have missed that file.

**Comment by wtrocki** [comment]
@rob3000 @krystianity   I have created discord server to talk about things https://discord.gg/H74peC

**Comment by wtrocki** [comment]
Link above will expire tommorow. Here you have new one: https://discord.gg/tnbQaFa

**Comment by wtrocki** [comment]
We can add this also to the repo readme at some point

**Comment by wtrocki** [comment]
> yarn lint
✖ 853 problems (393 errors, 460 warnings)

Linting is definitely separate issue :D

**Comment by wtrocki** [comment]
I think this is the first phase of the migration to typescript that can land into master.

What is outstanding:

- Move all public fields (currently as any) to the typed versions
- Address all linting issues (it will do first point as well).

@rob3000 thank you so much for amazing collab on this!

**Comment by rob3000** [comment]
solid work on this @wtrocki ! Happy to merge. I'll start looking to move the public and any types once this has been merged 😄

**Pull request merged**